### PR TITLE
Perceptual Similarity loss

### PR DESCRIPTION
**Pull request by tristan-deep**
As described in #20839. 

Conversion of weights partly based on an implementation found [here](https://github.com/moono/lpips-tf2.x/blob/master/models/lpips_tensorflow.py).

### Added
- [`keras.losses.PerceptualSimilarity`](https://github.com/tristan-deep/keras/blob/d74d3dd1095f43ec693f143f954a072e9671e2f2/keras/src/losses/losses.py#L1507)
- [`keras.applications.lpips.LPIPS`](https://github.com/tristan-deep/keras/blob/lpips/keras/src/applications/lpips.py)

### TODO
- add to Metrics?
- upload weights

I uploaded the LPIPS weights to [Hugging Face](https://huggingface.co/tristan-deep/lpips/blob/main/lpips_vgg16.weights.h5), as I'm not sure how to upload to `storage.googleapis.com`.

### Testing code snippet
Tests both the model and loss object and compares to the [torch metrics implementation](https://lightning.ai/docs/torchmetrics/stable/image/learned_perceptual_image_patch_similarity.html).

```python
# !pip install torch torchmetrics
# !pip install scikit-image

import torch
from skimage import data
from torchmetrics.image.lpip import LearnedPerceptualImagePatchSimilarity

from keras import ops
from keras.src.applications import lpips
from keras.src.losses import PerceptualSimilarity

## dummy images
image1 = data.camera()
image2 = data.gravel()

expected_lpips_value = 0.7237163782119751
print(f"Expected LPIPS value: {expected_lpips_value}")

image1 = image1[None, ..., None].astype("float32")
image2 = image2[None, ..., None].astype("float32")

image1 = ops.repeat(image1, 3, axis=-1)
image2 = ops.repeat(image2, 3, axis=-1)

# placeholder for weights
custom_weight_path = "/Users/tristanstevens/data/lpips.weights.h5"

## Keras model
model = lpips.LPIPS(weights=custom_weight_path)

model.summary()
print("Model loaded")

image1_tensor = lpips.preprocess_input(image1)
image2_tensor = lpips.preprocess_input(image2)

score = model([image1_tensor, image2_tensor])
print(f"Keras LPIPS score: {score}")

## Loss
loss = PerceptualSimilarity(weights=custom_weight_path)
value = loss(image1, image2)
print(f"Keras loss value: {value}")


## torchmetrics check
def preprocess_input_for_torch(x):
    # Convert from [0, 255] to [-1, 1]
    x = x / 127.5 - 1
    # Convert to numpy array
    x = ops.convert_to_numpy(x)
    # Rearrange from NHWC to NCHW format
    x = x.transpose(0, 3, 1, 2)
    # Convert to torch tensor
    x = torch.tensor(x)
    return x


image1_torch = preprocess_input_for_torch(image1)
image2_torch = preprocess_input_for_torch(image2)

# needs to be in range [-1, 1]
lpips = LearnedPerceptualImagePatchSimilarity(
    net_type="vgg",
    normalize=False,
)
score = lpips(image1_torch, image2_torch)

print(f"Torch LPIPS score: {score}")

```

### CODE REVIEW

File: keras/src/losses/losses.py [hunk 1]
```diff
@@ -1504,6 +1504,75 @@ def get_config(self):
...
+    def __init__(
+        self,
+        weights="imagenet",
+        network_type="vgg",
+        preprocess_inputs=True,
+        reduction="sum_over_batch_size",
+        name="lpips",
+        dtype=None,
+    ):
+        from keras.src.applications import lpips  # lazy import
```
Comment (tristan-deep) [inline keras/src/losses/losses.py:1555]: is there a better way to do this?

**Comment by tristan-deep** [comment]
@matt-gardner does it make more sense to leave LPIPS as an application and not include in the losses? That way people can always use it through a custom loss definition, without cluttering / redesigning the loss api for feature extraction losses.

**Comment by fchollet** [comment]
Thanks for the PR!

Due to the way this method works, it does not seem to be a good fit for the `keras.loss` API. In particular losses should not be importing Applications.

I think the best way to demonstrate this loss might be one of:

- A standalone github repo factored as a reusable template showing how to use LPIPS with any model
- A keras.io code example showing the same

The easiest way to leverage LPIPS might be to subclass a model and override the `compute_loss()` method.

**Comment by tristan-deep** [comment]
@fchollet Thanks for the quick look! I agree that currently it doesn't really fit in the loss api and have removed it. Do you think LPIPS is still useful as an application within Keras? In that case, how do we handle the upload of the weights here?

As a follow up PR for keras.io, a code example could demonstrate it in a training loop for instance as you suggest.

File: keras/src/applications/lpips.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
+from keras.src import backend
+from keras.src import layers
+from keras.src import ops
+from keras.src.api_export import keras_export
+from keras.src.applications import imagenet_utils
+from keras.src.applications import vgg16
+from keras.src.models import Functional
+from keras.src.utils import file_utils
+
+WEIGHTS_PATH = (
```
[the remark this answers is not in the document]
Comment (tristan-deep) [reply keras/src/applications/lpips.py:10]: Currently added test will fail because it cannot find the weights at that location. Easy way to try out the tests locally is to change this line:

https://github.com/keras-team/keras/blob/4709571dfc70dd186d2664af38e84b52c8cde542/keras/src/applications/applications_test.py#L284

with

```python
model = lpips.LPIPS(weights=<path_to_local_copy>)
```

**Comment by fchollet** [comment]
> Do you think LPIPS is still useful as an application within Keras? In that case, how do we handle the upload of the weights here?

We aren't adding any new models to the `keras.applications` API at the moment, since all new models are going to `keras_hub` instead. So I recommend not adding it there.
